### ISSUE TITLE
Attempting to write to a closed socket should raise an exception

### ISSUE THREAD

**Issue by oofnikj** [ronf/asyncssh#303]
First of all I'd like to preface this with the disclaimer that this could be entirely due to my poorly written code, but I think I've encountered an issue where an exception should be raised, but isn't.

I'm writing an interactive command shell using `Cmd.cmd` behind an AsyncSSH server. Normally one would call `Cmd.cmdloop()` to initialize the interactive shell, but since this method doesn't work with asyncio, I've approximated it with the following:
```py
async def shell(server, process):
    ns = NuttShell(server, process)  # subclass of cmd.Cmd hooked up to SSHServerProcess stdin and stdout
    ns.stdout.write(ns.intro)
    while True:
        try:
            ns.stdout.write(ns.prompt)
            line = (await ns.stdin.readline())
            if not line:
                raise BrokenPipeError
            ns.onecmd(line)
        except BrokenPipeError:
            break
        except (BreakReceived, misc.TerminalSizeChanged):
            ns.stdout.write('\n')
            pass
    process.exit(0)
```

This works reasonably well, with the exception of command completion, until someone does something nasty:
```sh
$ yes | ssh myserver.example.com
```

What follows is an endless response of `*** Unknown syntax: y`, which is expected. But the interesting part is what happens when the client closes the connection in the middle of the loop. There is now nowhere to `write()` to, and the following error is printed to the console, four times for each iteration of the loop, until `SSHServerProcess._recv_buf` is empty:
```
WARNING:asyncio:socket.send() raised exception.
```

In the mean time, the server remains unresponsive to new connections until the receive buffer has been completely flushed. This can take several minutes as the server has to work through several hundred thousand iterations of the above loop in the worst case.

The only reliable way I've managed to mitigate this is by adding a secondary condition to check the receive buffer length against some arbitrarily low value that should probably never be exceeded for an interactive shell:
```py
            if not line or process._recv_buf_len > 4096:
                raise BrokenPipeError
```
but this seems a bit hacky (even more so than the above `cmdloop()` replacement).

Several questions:
- What's the proper way of setting this receive buffer to something sane? I tried `process.channel.set_write_buffer_limits()`, but this didn't seem to have any effect.
- Is there a way to handle the `socket.send()` exception from asyncio so that we can immediately close the connection?

Thank you for your time.

**Comment by ronf** [comment]
Generally speaking, AsyncSSH (and asyncio in general, I think) relies on detecting that a socket connection has closed through the read side of the socket closing. As long as you are trying to read from the socket (or you have implemented the eof_received() method when using the low-level callback based async API), that should notify you when it's time to clean up. 

You don't show your implementation of onecmd(), but is that ever trying to read data from the process? If so, I would expect it to let you know when you've hit EOF (by returning an empty string on a read). Also, you should be able to test for EOF explicitly using the at_eof() call on your SSHReader() fields in SSHProcess.

I must admit I'm not sure I'm 100% following what you're trying to do here. How would you known when to be reading the next interactive shell command from the user vs. reading the output coming back from the last shell command the user entered?

**Comment by ronf** [comment]
I was able to reproduce this today, seeing the "WARNING:asyncio:socket.send() raised exception." messages generated by asyncio. I've seen those before on unexpected connection closes, so that's not a surprise, but it gave me a chance to take a closer look at some of the other things you found.

Based on my testing here, it would be fairly easy to add a check to reduce the change of generating the socket.send() warning messages. However, particularly with the test of piping "yes" directly to the SSH client, I found that a large amount of the processing of that input happened without the underlying asyncio code noticing the connection had gone away. So, even after putting in checks to try and signal connection lost based on write errors, it still continued processing input for minutes in some cases after I killed the SSH client connection, and so the result wasn't better than the checks already in place that trigger a connection lost based on the read side of the TCP transport detecting the connection close.

Even if you try to exit early in your application code, I think you are leveraging the line-editor code to echo the user input when your application calls readline() in this case, and all of that processing still happens on all the buffered input read before the connection is closed. I could try and add a check there to disable line editing if the transport is closing, but the editor code really shouldn't know anything about the transport connection.

More specifically, it looks like the data being sent comes in as 16 KB reads, so that would be 8192 lines of input to process in the line editor (and eventually send through to the application), all of which would have arrived before the socket was closed. Even if the socket send failed and printed the WARNING message while generating command responses, that would only abort out of the application code early, but the editor processing would all still happen.

Also, getting a socket write error doesn't seem to set the _closing flag on the underlying asyncio transport. It sets something internal called _conn_lost, but I wouldn't want to rely on that.

**Comment by ronf** [comment]
To summarize, here's what I think is going on:

- When a write error occurs on a socket in asyncio, it doesn't return an indication of that to the caller directly, but instead triggers a call_soon() to the protocol's connection_lost() method. AsyncSSH already implements this to trigger closing down of all the channels on a connections (and eventually the SSHConnection itself). An application can also register a global exception handler to be called, but that's not something AsyncSSH can do, since it is just a library and you can only have one global exception handler for all I/O. That's ok, though, since the connection_lost() callback is just fine for AsyncSSH.

- The socket.send warnings can be generated when more than 5 write() calls are made after the connection is lost. This can happen when AsyncSSH is still processing input which arrived before the connection was lost, since that code will run before the connection_lost() callback is scheduled by asyncio.

- As soon as the connection_lost() is run on the connection, any remaining data in channel output buffers is flushed by AsyncSSH and any future sends will trigger a BrokenPIpeError, which should allow your application to exit before processing all of the remaining queued input. However, the exact timing of this depends on when asyncio schedules the connection_lost() callback compared with data_received() callbacks for any data sent before the connection was lost.

**Comment by oofnikj** [comment]
Hi, thanks for taking a more in-depth look at this.

Makes sense re: the half-open TCP state, and I agree that in most cases it's probably fine to continue writing to the output buffer even if the socket enters the closing state to make sure connection closures are handled cleanly. I was just trying to handle this particular case of someone flooding the interactive line editor with junk input.

For this particular case, I think the harm in raising a `BrokenPipeError` in the event that the socket begins closing while inside the interactive shell loop is outweighed by the benefit. It's a simple way to mitigate the effects of what amounts to a DOS attack.

For now I've decided to leave both safeguards in - check that receive buffer length has not exceeded 4096 bytes, and check that the transport is not in the closing state. If either condition is true inside the loop, raise a `BrokenPipeError` exception.

**Comment by ronf** [comment]
As part of AsyncSSH 2.5.0, I've included commit dc353df, which does an explicit check for is_closing() on any attempt to write to the transport used by AsyncSSH and raises a BrokenPipeError even if the transport isn't fully closed yet. This should allow applications to receive the BrokenPipeError more quickly than they would have in prior releases.

I've also made substantial performance improvements in the AsyncSSH line editor code in this release, and added in a default maximum line length of 1024 bytes when the line editor is active (though you can raise that or even set it back to unlimited if you want via the max_line_length option). This should help to protect against denial-of-service attacks.
